### DISCUSSION TITLE
Errors Initiating connection to subscription service: It was most likely caused due to an unsupported subprotocol "graphql-ws" requested by the client.graphql-ws implements exclusively the "graphql-transport-ws" subprotocol, please make sure that the client implements it too

### DISCUSSION THREAD

**Discussion by Efosaok** [enisdenjo/graphql-ws#237] [Q&A] [answered]
**Expected Behaviour**
Following an upgrade of Apollo Server Express from version 2 to 3 using Nestjs. It was recommended to use Graphql-ws to keep Apollo subscription related events functional. However, I ran into issues trying to connect the client(now set up to use Graphql-ws) to the server. The error stack is as below;

```
WebSocket protocol error occured. It was most likely caused due to an unsupported subprotocol "graphql-ws" requested by the client. graphql-ws implements exclusively the "graphql-transport-ws" subprotocol, please make sure that the client implements it too.
```

**Debug Information**
__Client Setup__

`wsLink.ts`

```
import { ApolloLink, FetchResult, Observable, Operation } from '@apollo/client/core';
import { GraphQLError, print } from 'graphql';
import { Client, ClientOptions, createClient } from 'graphql-ws';

export class WebSocketLink extends ApolloLink {
  private client: Client;

  constructor(options: ClientOptions) {
    super();
    this.client = createClient(options);
  }

  public request(operation: Operation): Observable<FetchResult> {
    return new Observable((sink) => {
      return this.client.subscribe<FetchResult>(
        { ...operation, query: print(operation.query) },
        {
          next: sink.next.bind(sink),
          complete: sink.complete.bind(sink),
          error: (err) => {
            if (err instanceof Error) {
              return sink.error(err);
            }

            if (err instanceof CloseEvent) {
              return sink.error(
                // reason will be available on clean closes
                new Error(`Socket closed with event ${err.code} ${err.reason || ''}`),
              );
            }

            return sink.error(
              new Error((err as GraphQLError[]).map(({ message }) => message).join(', ')),
            );
          },
        },
      );
    });
  }
}

const wsLink = new WebSocketLink({
  url: socketUrl,
  lazyCloseTimeout: 50000,
  retryAttempts: Infinity,
  lazy: true,
  connectionParams: () => {
    const token = sessionStorage.getItem('key');
    return {
      authorization: token ? `Bearer ${token}` : '',
    };
  },
});

```

**Further Information**
Installed version of `graphql-ws` is `5.5.0`.
I have tried downgrading to the `graphql-transport-ws` package but the error still persist.

**Comment by enisdenjo** [comment, marked as answer]
Hey there, as the server error suggest, your client is somewhere still using the [`subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws). No need to downgrade, the `graphql-transport-ws` subprotocol is used in all versions of this library - please update to the latest version.

Can you share (or check) more of your client-side code? Is this the only client? Do you use any GraphQL playground tools (like GraphQL Playground, Altair or GraphiQL)? You must have a `subscriptions-transport-ws` instance lurking somewhere.

**Reply by Efosaok** [reply]
I have since upgraded. I just added that bit of information to provide more context on the kind of troubleshooting I have tried. I believe I have totally removed `subscriptions-transport-ws` from the client because the `wsLink` above is the only instance that attempts to establish a connection in the entire app.

**Reply by enisdenjo** [reply]
I think the error is coming from a rogue client or GraphQL UI tool. The only client that can request a [`"graphql-ws"` is `subscriptions-transport-ws`](https://github.com/apollographql/subscriptions-transport-ws/blob/231c1b0d25b20090033140b83c8a39871d4ed88c/src/protocol.ts#L1). This library has been [requesting a `"graphql-transport-ws"` subprotocol since its inception](https://github.com/enisdenjo/graphql-ws/blob/43da515de8e42c30f045cb02a1bc979f2c6b1aa9/src/common.ts#L16-L21), it simply cannot request another one.

You can confirm that by adding a few console logs. Alter your client like this:

```diff
const wsLink = new WebSocketLink({
  url: socketUrl,
  lazyCloseTimeout: 50000,
  retryAttempts: Infinity,
-  lazy: true,
+  lazy: false, // make the client connect immediately
+  on: {
+    connected: () => console.log('graphql-ws connected'),
+    error: (err) => console.log(err),
+  },
  connectionParams: () => {
    const token = sessionStorage.getItem('key');
    return {
      authorization: token ? `Bearer ${token}` : '',
    };
  },
});
```

If the DevTools of your browser logs `"graphql-ws connected"` (without a following `"graphql-ws failed"` error), the `graphql-ws` client has successfully connected and you have somewhere an another, separate, client running `subscriptions-transport-ws` in your environment trying to connect.

**Reply by Efosaok** [reply]
Thanks for the pointer @enisdenjo . apparently I had multiple clients opened in different tabs that had not reloaded to use the new protocol so those clients were trying to initiate the connection with the previous setup. killing them solved it.

Thanks.

**Reply by enisdenjo** [reply]
Awesome, glad you figured it out! 😄

**Reply by mrparkersson** [reply]
@enisdenjo I have implemented this `subscriptions: {
        'graphql-ws': true,
        'subscriptions-transport-ws': true,
      },` on my Nest Js and I get **undefined** when I try making a request. Do you have any idea or how to solve this?

**Reply by enisdenjo** [reply]
@mrparkersson I really don't know - am not a user of NestJS. You could ask them instead.

**Reply by mrparkersson** [reply]
@enisdenjo sure brother thank you. :)

**Reply by enisdenjo** [reply]
@mrparkersson you're very welcome, sorry I couldn't be of help.

**Reply by mrparkersson** [reply]
@enisdenjo  oh don't worry. We all don't know everything
